File: webapp/codigos-fonte/pilha.c
#include <stdio.h>
#include <stdlib.h>

struct no {
    int valor;
    struct no *proximo;
};

typedef struct no No;

struct pilha{
    No *topo;
};

typedef struct pilha Pilha;

Pilha *criarPilha() {
    Pilha *p = (Pilha *) malloc(sizeof(Pilha));
    p->topo = NULL;
}

void empilhar(Pilha *p, int valor){
    No *novoNo = (No *) malloc(sizeof(No));
    novoNo->valor = valor;
    novoNo->proximo = p->topo;
    p->topo = novoNo;
}

void desempilhar(Pilha *p){
    if(p->topo == NULL){
        printf("Erro ao tentar retirar um elemento de uma pilha vazia\n");
    }
    No *temporario = p->topo;
    p->topo = p->topo->proximo;
    int valor = temporario->valor;
    free(temporario);
}

void printAll(Pilha *p){
    No *atual = p->topo;
    while(atual != NULL){
        printf("%d ", atual->valor);
        atual = atual->proximo;
    }
    printf("\n");
}

int main() {
    Pilha *p = criarPilha();
    for(int i = 0; i < 10; i++){
        empilhar(p, i);
        printAll(p);
    }
    for(int i = 0; i < 10; i++){
        desempilhar(p);
        printAll(p);
    }
}

File: webapp/codigos-fonte/Kotlin/pilha.kt
package pilha

data class No (var valor:Int, var proximo: No? = null)

data class Pilha (var topo:No? = null) {
    fun empilhar(valor:Int) {
        val novoNo = No(valor)
        novoNo.proximo = this.topo
        this.topo = novoNo
    }

    fun desempilhar() {
        this.topo?: let {
            println("Erro ao tentar retirar um elemento de uma pilha vazia");
            return
        }
        this.topo = this.topo?.proximo
    }

    fun printAll() {
        var atual = this.topo
        while (atual != null) {
            print(String.format("%d ", atual.valor))
            atual = atual.proximo
        }
        println()
    }
}

fun main() {
    val p = Pilha();
    p.empilhar(5)
    p.printAll()
    p.empilhar(10)
    p.printAll()
    p.empilhar(15)
    p.printAll()
    p.empilhar(20)
    p.printAll()
    p.desempilhar()
    p.printAll()
    p.desempilhar()
    p.printAll()
}